File: backend/app/schemas/overlord.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.models import EventSeverity

OverlordSystemStatus = Literal[
    "healthy",
    "degraded",
    "blocked",
    "rate-limited",
    "operator-required",
    "auto-recovering",
    "stale-state-mismatch",
]
OverlordIncidentStatus = Literal["active", "resolved"]
OverlordRemediationAction = Literal[
    "safe_pause",
    "safe_resume",
    "safe_retry",
    "stale_state_cleanup",
    "resolved_alert_cleanup",
    "cooldown_aware_retry_scheduling",
    "notify",
]


class OverlordIncident(BaseModel):
    incident_key: str
    title: str
    status: OverlordIncidentStatus = "active"
    system_status: OverlordSystemStatus
    severity: EventSeverity
    summary: str
    agent_name: str | None = None
    provider: str | None = None
    detected_at: datetime | None = None
    last_observed_at: datetime | None = None
    retry_after_seconds: int | None = None
    requires_operator: bool = False
    auto_recovering: bool = False
    why_it_happened: str
    what_overlord_did: str | None = None
    operator_action: str | None = None


class OverlordActionRecord(BaseModel):
    action: OverlordRemediationAction
    target: str
    summary: str
    created_at: datetime | None = None
    status: Literal["applied", "skipped", "resolved"]


class OverlordTelegramStatus(BaseModel):
    enabled: bool
    min_severity: EventSeverity = EventSeverity.ERROR
    daily_digest_enabled: bool = False
    configured_chat: bool = False
    configured_token: bool = False


class OverlordSummaryResponse(BaseModel):
    agent_name: str = "overlord"
    display_name: str = "Overlord"
    status: OverlordSystemStatus
    headline: str
    summary: str
    generated_at: datetime
    incidents: list[OverlordIncident] = Field(default_factory=list)
    recent_actions: list[OverlordActionRecord] = Field(default_factory=list)
    operator_todos: list[str] = Field(default_factory=list)
    telemetry: dict[str, int | str | bool | None] = Field(default_factory=dict)
    telegram: OverlordTelegramStatus


class OverlordPolicyResponse(BaseModel):
    auto_remediation_enabled: bool
    safe_pause_enabled: bool
    safe_resume_enabled: bool
    stale_state_cleanup_enabled: bool
    telegram: OverlordTelegramStatus

File: frontend/src/api/overlord.ts
import { getRequiredApiBaseUrl } from "./base-url";
import type { EventSeverity } from "./agents";

export type OverlordSystemStatus =
  | "healthy"
  | "degraded"
  | "blocked"
  | "rate-limited"
  | "operator-required"
  | "auto-recovering"
  | "stale-state-mismatch";

export interface OverlordIncident {
  incident_key: string;
  title: string;
  status: "active" | "resolved";
  system_status: OverlordSystemStatus;
  severity: EventSeverity;
  summary: string;
  agent_name: string | null;
  provider: string | null;
  detected_at: string | null;
  last_observed_at: string | null;
  retry_after_seconds: number | null;
  requires_operator: boolean;
  auto_recovering: boolean;
  why_it_happened: string;
  what_overlord_did: string | null;
  operator_action: string | null;
}

export interface OverlordActionRecord {
  action: string;
  target: string;
  summary: string;
  created_at: string | null;
  status: "applied" | "skipped" | "resolved";
}

export interface OverlordTelegramStatus {
  enabled: boolean;
  min_severity: EventSeverity;
  daily_digest_enabled: boolean;
  configured_chat: boolean;
  configured_token: boolean;
}

export interface OverlordSummaryResponse {
  agent_name: "overlord";
  display_name: "Overlord";
  status: OverlordSystemStatus;
  headline: string;
  summary: string;
  generated_at: string;
  incidents: OverlordIncident[];
  recent_actions: OverlordActionRecord[];
  operator_todos: string[];
  telemetry: Record<string, string | number | boolean | null>;
  telegram: OverlordTelegramStatus;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isOverlordSummaryResponse(value: unknown): value is OverlordSummaryResponse {
  return (
    isRecord(value) &&
    value.agent_name === "overlord" &&
    typeof value.display_name === "string" &&
    typeof value.status === "string" &&
    typeof value.headline === "string" &&
    typeof value.summary === "string" &&
    Array.isArray(value.incidents) &&
    Array.isArray(value.recent_actions) &&
    Array.isArray(value.operator_todos) &&
    isRecord(value.telemetry) &&
    isRecord(value.telegram)
  );
}

export function getOverlordSummaryQueryKey() {
  return ["overlord", "summary"] as const;
}

export async function fetchOverlordSummary(): Promise<OverlordSummaryResponse> {
  const response = await fetch(`${getRequiredApiBaseUrl()}/api/v1/overlord/summary`, {
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error(`Failed to fetch Overlord summary: ${response.status} ${response.statusText}`);
  }
  const data: unknown = await response.json();
  if (!isOverlordSummaryResponse(data)) {
    throw new Error("Overlord summary response has unexpected shape");
  }
  return data;
}
